modules/thermal_hydraulics/src/base/Simulation.C

```diff
@@ -564,6 +572,63 @@ Simulation::addComponentScalarIC(const VariableName & var_name, const std::vecto
   addSimInitialCondition(class_name, genName(var_name, "ic"), params);
 }
 
+std::vector<VariableName>
+Simulation::sortAddedComponentVariables() const
+{
+  // Check that no index in order map is used more than once.
+  // Also, convert the map to a vector pairs to be sorted.
+  std::vector<int> indices;
+  std::vector<std::pair<VariableName, int>> registered_var_index_pairs;
+  for (const auto & var_and_index : _component_variable_order_map)
+  {
+    registered_var_index_pairs.push_back(var_and_index);
+
+    const auto ind = var_and_index.second;
+    if (std::find(indices.begin(), indices.end(), ind) != indices.end())
+      mooseError("The index ", ind, " was used for multiple component variables.");
+    else
+      indices.push_back(ind);
+  }
+
+  // Collect all of the added variable names into an unsorted vector.
+  std::vector<VariableName> vars_unsorted;
+  for (const auto & var_and_data : _vars)
+    vars_unsorted.push_back(var_and_data.first);
+
+  // The sorting works as follows. For those variables that are listed in
+  // _component_variable_order_map, these are ordered before those that are not,
+  // in the order of their indices in the map. Those not in the map are sorted
+  // alphabetically.
+
+  // Sort registered_var_index_pairs by value (index)
+  std::sort(registered_var_index_pairs.begin(),
+            registered_var_index_pairs.end(),
+            [](const std::pair<VariableName, int> & a, const std::pair<VariableName, int> & b)
+            { return a.second < b.second; });
+
+  // Loop over the ordered, registered variable names and add a variable to the
+  // sorted list if in vars_unsorted. When this happens, delete the element from
+  // vars_unsorted, leaving only unregistered variable names after the loop.
+  std::vector<VariableName> vars_sorted;
+  for (const auto & var_index_pair : registered_var_index_pairs)
+  {
+    const auto & var = var_index_pair.first;
+    if (std::find(vars_unsorted.begin(), vars_unsorted.end(), var) != vars_unsorted.end())
```

Comment:
if there's no duplicates do you really need to search? you can always insert in vars_sorted and delete in vars_unsorted